File: src/core/Layout.ts
/**
 * Single source of truth for screen + zone geometry.
 *
 * Plain numbers, Phaser-free, so both halves and the tests can read it without
 * booting the engine. Both devs read this; it changes rarely and only by mutual
 * agreement (it sits behind the seam).
 */

/** Design resolution — locked portrait. The Scale manager fits this to the device. */
export const WIDTH = 390;
export const HEIGHT = 844;

export interface Rect {
  x: number;
  y: number;
  width: number;
  height: number;
}

// Vertical split (tunable). Zone A = big merge board on top, Zone C = thin
// trap-door band, Zone B = pinball arena below.
//
// The three zones NO LONGER tile the 844px screen: the world is 1238px tall and the
// game alternates between two camera framings (see core/phaseGeometry.ts):
//  - A-phase (scroll 0): the full Zone A band is on screen — 2/3 of the screen
//    (42px HUD + 521px board = 563 = round(844 × 2/3)); Zone B is bottom-cropped by 394px.
//  - B-phase (scroll 394): Zone A shows only its bottom 1/5 of the screen (42px HUD +
//    127px board = 169 = round(844 / 5), top-cropped); Zone B's full 631px band exactly
//    fills the screen down to y=844.
// Zone A's band: 42 (HUD chrome) + 521, where 42 + 521 = round(844 × 2/3).
// Zone B's band: 844 − 42 − 127 − 44 = 631, where 42 + 127 = round(844 / 5).
const ZONE_A_HEIGHT = 42 + 521;
const ZONE_C_HEIGHT = 44;
const ZONE_B_HEIGHT = 631;

export const zoneA: Rect = { x: 0, y: 0, width: WIDTH, height: ZONE_A_HEIGHT };
export const zoneC: Rect = { x: 0, y: zoneA.y + zoneA.height, width: WIDTH, height: ZONE_C_HEIGHT };
export const zoneB: Rect = { x: 0, y: zoneC.y + zoneC.height, width: WIDTH, height: ZONE_B_HEIGHT };

/**
 * The fixed column every ball enters Zone B through (top-center of Zone B).
 *
 * FROZEN: this is the `x` Zone C stamps into every `BALL_DROPPED` payload. See the
 * note on `BallDroppedPayload.x` in contracts.ts.
 */
export const zoneBEntry = { x: WIDTH / 2, y: zoneB.y };

File: Assets/Game/Core/Layout.cs
using UnityEngine;

namespace RichCoast.Core
{
    /// <summary>Axis-aligned rectangle in design space (y grows downward, as in the original design).</summary>
    public readonly struct ZoneRect
    {
        public readonly float X;
        public readonly float Y;
        public readonly float Width;
        public readonly float Height;

        public ZoneRect(float x, float y, float width, float height)
        {
            X = x;
            Y = y;
            Width = width;
            Height = height;
        }

        public float Right => X + Width;
        public float Bottom => Y + Height;
        public float CenterX => X + Width * 0.5f;
    }

    /// <summary>
    /// Single source of truth for world + zone geometry. Plain numbers, no scene dependency, so
    /// every zone and the tests read it without booting anything.
    ///
    /// Design space is 390 wide; the WORLD is 1238 tall — deliberately taller than the ~844 the
    /// screen shows — and the game alternates between two camera framings rather than ever
    /// moving world objects (see <see cref="PhaseGeometry"/>):
    ///
    ///  - A-phase: the full Zone A band is on screen (~60% of the screen: 42 HUD + 465 board);
    ///    Zone B is bottom-cropped.
    ///  - B-phase: Zone A shows only a 113 sliver (42 HUD + 71 board, top-cropped) and Zone B's
    ///    full 687 band exactly fills the rest of the screen.
    ///
    /// Zone A's band is round(844 × 2/3 × 0.9) — the original 2/3 split, shrunk 10% in Zone B's
    /// favour.
    ///
    /// Device fitting: width is locked to <see cref="Width"/> and the visible height follows the
    /// device aspect ratio (<see cref="ScreenHeightForAspect"/>), so taller or shorter phones get
    /// more or less of the world rather than being letterboxed. The original build assumed a
    /// fixed 390×844 screen.
    /// </summary>
    public static class Layout
    {
        /// <summary>Design width — locked portrait. World units are design pixels.</summary>
        public const float Width = 390f;

        /// <summary>The screen height the design was authored against (390×844, a 2019-era phone).</summary>
        public const float DesignScreenHeight = 844f;

        private const float ZoneAHeight = 42f + 465f;
        private const float ZoneCHeight = 44f;
        private const float ZoneBHeight = 687f;

        public static readonly ZoneRect ZoneA = new ZoneRect(0f, 0f, Width, ZoneAHeight);
        public static readonly ZoneRect ZoneC = new ZoneRect(0f, ZoneA.Bottom, Width, ZoneCHeight);
        public static readonly ZoneRect ZoneB = new ZoneRect(0f, ZoneC.Bottom, Width, ZoneBHeight);

        /// <summary>Total world height: the three bands stacked (1238).</summary>
        public static float WorldHeight => ZoneB.Bottom;

        /// <summary>
        /// Visible design-space height for a device aspect ratio, given the locked design width.
        /// A 390×844 phone yields exactly <see cref="DesignScreenHeight"/>; a taller phone shows
        /// more world, a shorter one less.
        /// </summary>
        public static float ScreenHeightForAspect(float widthOverHeight)
        {
            if (widthOverHeight <= 0f) return DesignScreenHeight;
            return Width / widthOverHeight;
        }

        /// <summary>Convenience for runtime code: the visible height on the current screen.</summary>
        public static float CurrentScreenHeight()
        {
            var w = Mathf.Max(1, Screen.width);
            var h = Mathf.Max(1, Screen.height);
            return ScreenHeightForAspect((float)w / h);
        }
    }
}
